src/main/java/com/example/petbuddybackend/service/image/FirebasePhotoService.java

```diff
@@ -0,0 +1,157 @@
+import com.example.petbuddybackend.entity.photo.CloudPhoto;
+import com.example.petbuddybackend.repository.photo.CloudPhotoRepository;
+import com.example.petbuddybackend.utils.exception.throweable.general.NotFoundException;
+import com.example.petbuddybackend.utils.exception.throweable.photo.InvalidPhotoException;
+import com.google.cloud.storage.Blob;
+import com.google.cloud.storage.Bucket;
+import com.google.firebase.FirebaseApp;
+import com.google.firebase.cloud.StorageClient;
+import lombok.RequiredArgsConstructor;
+import org.apache.tika.Tika;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.stereotype.Service;
+import org.springframework.web.multipart.MultipartFile;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.time.LocalDateTime;
+import java.util.*;
+import java.util.concurrent.TimeUnit;
+
+
+@Service
+@RequiredArgsConstructor
+public class FirebasePhotoService implements PhotoService {
+
+    private final static String INVALID_PHOTO_PROVIDED_MESSAGE = "Invalid photo provided";
+    private final static String MISSING_FILE_EXTENSION = "Missing file extension";
+    private final static String PHOTO = "Photo";
+    private static final Set<String> acceptedImageTypes = Set.of(
+            "image/jpeg",
+            "image/jpg",
+            "image/png",
+            "image/webp"
+    );
+
+    @Value("${firebase.photo.directory}")
+    private String PHOTO_DIRECTORY;
+
+    @Value("${firebase.photo.expiration.max-seconds}")
+    private Integer MAX_EXPIRATION_SECONDS;
+
+    @Value("${firebase.photo.expiration.threshold-seconds}")
+    private Integer EXPIRATION_THRESHOLD_SECONDS;
+
+    private final FirebaseApp firebaseApp;
+    private final CloudPhotoRepository cloudPhotoRepository;
+    private final Tika tika;
+
+
+    @Override
+    public CloudPhoto getPhoto(String blob) {
+        return cloudPhotoRepository.findById(blob)
+                .orElseThrow(() -> NotFoundException.withFormattedMessage(PHOTO, blob));
+    }
+
+    @Override
+    public CloudPhoto uploadPhoto(MultipartFile multipartFile) {
+        validatePhoto(multipartFile);
+        CloudPhoto photo = uploadFile(multipartFile, MAX_EXPIRATION_SECONDS);
+        return cloudPhotoRepository.save(photo);
+    }
+
+    @Override
+    public void deletePhoto(String blob) {
+        CloudPhoto photo = getPhoto(blob);
+        StorageClient storageClient = StorageClient.getInstance(firebaseApp);
+        Bucket bucket = storageClient.bucket();
+        Blob blobToDelete = bucket.get(blob);
+
+        if(blobToDelete != null) {
+            blobToDelete.delete();
+        }
+
+        cloudPhotoRepository.delete(photo);
+    }
+
+    @Override
+    public CloudPhoto updatePhotoExpiration(CloudPhoto photo) {
+        LocalDateTime thresholdTime = photo
+                .getUrlExpiresAt()
+                .minusSeconds(EXPIRATION_THRESHOLD_SECONDS);
+
+        if(thresholdTime.isAfter(LocalDateTime.now())) {
```

Comment:
czemu ustawiamy ograniczenie, ze mozna odnowic tylko zdjecie bliskie wygasniecia?

Reply:
Pomyślałem, że nie trzeba odnawiać co każdy call do backendu, tylko jak już jest bliskie wygaśnięciu

Reply:
W sumie można by było odświeżyć w trakcie getPhoto. Albo można rozważyć scheduled i co jakiś czas odświeżać, ale wtedy równie dobrze można ustawić nieskończoność.

Reply:
> W sumie można by było odświeżyć w trakcie getPhoto

W sumie czemu nie


> Albo można rozważyć scheduled i co jakiś czas odświeżać, ale wtedy równie dobrze można ustawić nieskończoność

Raczej nie ma potrzeby takiej. Tak jak jest teraz to byłoby lazy odświeżanie, czyli zdjęcie się odświeży w momencie gdy będzie potrzebne